File: server-shared/src/test/java/rikka/shizuku/server/ClientManagerTest.java
package rikka.shizuku.server;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertThrows;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static rikka.shizuku.server.ServiceCallerGateTest.application;
import static rikka.shizuku.server.ServiceCallerGateTest.entry;

import android.os.IBinder;
import android.os.RemoteException;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.util.List;

import moe.shizuku.server.IShizukuApplication;

/** What {@link ClientManager} records, how it seeds {@code allowed}, and when it forgets a record. */
@RunWith(RobolectricTestRunner.class)
@Config(sdk = 34, manifest = Config.NONE)
public class ClientManagerTest {

    private static final String PACKAGE = "eu.darken.porter.probe";
    private static final int UID = 10200;
    private static final int PID = 45678;

    private ConfigManager config;
    private ClientManager<ConfigManager> clients;

    @Before
    public void setup() {
        config = mock(ConfigManager.class);
        clients = new ClientManager<>(config);
    }

    private ClientRecord add(int uid, int pid, IShizukuApplication application) {
        return clients.addClient(uid, pid, application, PACKAGE, 13);
    }

    @Test
    public void addClientSeedsAllowedFromAnAllowedEntry() {
        when(config.find(UID)).thenReturn(entry(true, false));

        ClientRecord record = add(UID, PID, application(mock(IBinder.class)));

        assertNotNull(record);
        assertTrue(record.allowed);
    }

    @Test
    public void addClientLeavesAllowedFalseForNoEntryOrDeniedEntry() {
        assertFalse(add(UID, PID, application(mock(IBinder.class))).allowed);

        when(config.find(UID)).thenReturn(entry(false, true));

        assertFalse(add(UID, PID + 1, application(mock(IBinder.class))).allowed);
    }

    @Test
    public void addClientLinksToDeathAndDeathForgetsTheRecord() throws Exception {
        IBinder binder = mock(IBinder.class);
        ClientRecord record = add(UID, PID, application(binder));
        assertSame(record, clients.findClient(UID, PID));

        ArgumentCaptor<IBinder.DeathRecipient> recipient =
                ArgumentCaptor.forClass(IBinder.DeathRecipient.class);
        verify(binder).linkToDeath(recipient.capture(), eq(0));
        recipient.getValue().binderDied();

        assertNull(clients.findClient(UID, PID));
    }

    @Test
    public void addClientReturnsNullWhenLinkToDeathFails() throws Exception {
        IBinder binder = mock(IBinder.class);
        doThrow(new RemoteException("dead")).when(binder).linkToDeath(any(IBinder.DeathRecipient.class), anyInt());

        assertNull(add(UID, PID, application(binder)));
        assertNull(clients.findClient(UID, PID));
    }

    @Test
    public void findClientsReturnsEveryRecordOfTheUid() {
        ClientRecord first = add(UID, PID, application(mock(IBinder.class)));
        ClientRecord second = add(UID, PID + 1, application(mock(IBinder.class)));
        ClientRecord other = add(UID + 1, PID + 2, application(mock(IBinder.class)));

        assertEquals(List.of(first, second), clients.findClients(UID));
        assertEquals(List.of(other), clients.findClients(UID + 1));
    }

    @Test
    public void requireClientThrowsIllegalStateWhenAbsent() {
        assertThrows(IllegalStateException.class, () -> clients.requireClient(UID, PID));
    }

    @Test
    public void requireClientWithPermissionThrowsSecurityWhenNotAllowed() {
        ClientRecord record = add(UID, PID, application(mock(IBinder.class)));

        assertSame(record, clients.requireClient(UID, PID, false));
        assertThrows(SecurityException.class, () -> clients.requireClient(UID, PID, true));

        record.allowed = true;

        assertSame(record, clients.requireClient(UID, PID, true));
    }
}

File: server-shared/src/test/java/rikka/shizuku/server/ClientManagerTest.kt
package rikka.shizuku.server

import android.os.IBinder
import android.os.RemoteException
import moe.shizuku.server.IShizukuApplication
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertSame
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mockito.doThrow
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import rikka.shizuku.server.ServerTestSupport.application
import rikka.shizuku.server.ServerTestSupport.entry

/** What [ClientManager] records, how it seeds `allowed`, and when it forgets a record. */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34], manifest = Config.NONE)
class ClientManagerTest {

    private lateinit var config: ConfigManager
    private lateinit var clients: ClientManager<ConfigManager>

    @Before
    fun setup() {
        config = mock(ConfigManager::class.java)
        clients = ClientManager(config)
    }

    private fun add(uid: Int, pid: Int, application: IShizukuApplication): ClientRecord? =
        clients.addClient(uid, pid, application, PACKAGE, 13)

    @Test
    fun addClientSeedsAllowedFromAnAllowedEntry() {
        `when`(config.find(UID)).thenReturn(entry(true, false))

        val record = add(UID, PID, application(mock(IBinder::class.java)))

        assertNotNull(record)
        assertTrue(record!!.allowed)
    }

    @Test
    fun addClientLeavesAllowedFalseForNoEntryOrDeniedEntry() {
        assertFalse(add(UID, PID, application(mock(IBinder::class.java)))!!.allowed)

        `when`(config.find(UID)).thenReturn(entry(false, true))

        assertFalse(add(UID, PID + 1, application(mock(IBinder::class.java)))!!.allowed)
    }

    @Test
    fun addClientLinksToDeathAndDeathForgetsTheRecord() {
        val binder = mock(IBinder::class.java)
        val record = add(UID, PID, application(binder))
        assertSame(record, clients.findClient(UID, PID))

        val recipient = ArgumentCaptor.forClass(IBinder.DeathRecipient::class.java)
        verify(binder).linkToDeath(recipient.capture(), eq(0))
        recipient.value.binderDied()

        assertNull(clients.findClient(UID, PID))
    }

    @Test
    fun addClientReturnsNullWhenLinkToDeathFails() {
        val binder = mock(IBinder::class.java)
        doThrow(RemoteException("dead")).`when`(binder).linkToDeath(any(IBinder.DeathRecipient::class.java), anyInt())

        assertNull(add(UID, PID, application(binder)))
        assertNull(clients.findClient(UID, PID))
    }

    @Test
    fun findClientsReturnsEveryRecordOfTheUid() {
        val first = add(UID, PID, application(mock(IBinder::class.java)))
        val second = add(UID, PID + 1, application(mock(IBinder::class.java)))
        val other = add(UID + 1, PID + 2, application(mock(IBinder::class.java)))

        assertEquals(listOf(first, second), clients.findClients(UID))
        assertEquals(listOf(other), clients.findClients(UID + 1))
    }

    @Test
    fun requireClientThrowsIllegalStateWhenAbsent() {
        assertThrows(IllegalStateException::class.java) { clients.requireClient(UID, PID) }
    }

    @Test
    fun requireClientWithPermissionThrowsSecurityWhenNotAllowed() {
        val record = add(UID, PID, application(mock(IBinder::class.java)))!!

        assertSame(record, clients.requireClient(UID, PID, false))
        assertThrows(SecurityException::class.java) { clients.requireClient(UID, PID, true) }

        record.allowed = true

        assertSame(record, clients.requireClient(UID, PID, true))
    }

    private companion object {
        const val PACKAGE = "eu.darken.porter.probe"
        const val UID = 10200
        const val PID = 45678
    }
}
